Add support for public keys encoded with PKCS#1

RonEld opened the pull request:
## Description
1) Add support for public keys encoded with PKCS#1
2) Add tests for PKCS#1 PEM and DER, and PKCS#8 DER

This PR should resolve #1122 

## Status
**READY**

## Requires Backporting
NO  
This is an enhancement, no need for backporting

## Additional comments
Added tests for parsing PKCS#1 public key in PEM and DER format

## Todos
- [x] Tests
- [ ] Documentation
- [x] Changelog updated
- [ ] Backported

hanno-arm commented on library/pkparse.c:
```diff
@@ -666,52 +675,77 @@ static int pk_parse_key_pkcs1_der( mbedtls_rsa_context *rsa,
+     *                   publicExponent    INTEGER   -- e
+     *                   }
+     */
+        if( ( ret = mbedtls_asn1_get_mpi( &p, end, &rsa->N  ) ) != 0 ||
```
We do already have code for this in `pk_get_rsapubkey`. I would prefer a solution which avoids this duplication.

hanno-arm commented on library/pkparse.c:
```diff
@@ -1255,12 +1289,43 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+         if( ( ret = mbedtls_pk_setup( ctx, pk_info ) ) != 0 )
+               return( ret );
+
+         if( ( ret = pk_parse_key_pkcs1_der( mbedtls_pk_rsa( *ctx ),
```
You could call the existing `pk_get_rsapubkey` from here.

RonEld replied:
I tried, `pk_get_rsapubkey` needs start and end, which are calculated in `pk_parse_key_pkcs1_der`

hanno-arm replied:
@RonEld You can setup a local variable holding `pem.buf` initially and pass its address as `p`, and `pem.buf + pem.buflen` as `end`.

RonEld replied:
@hanno-arm yes, I know, but I prefered not adding too much variables. If it's a blocker, I'll do it

hanno-arm commented on library/pkparse.c:
```diff
@@ -637,19 +637,28 @@ int mbedtls_pk_parse_subpubkey( unsigned char **p, const unsigned char *end,
+    if( ( ret = mbedtls_asn1_get_tag( &p, end, &len,
+            MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE ) ) != 0 )
+    {
+        return( MBEDTLS_ERR_PK_KEY_INVALID_FORMAT + ret );
```
We seem to have a specific error code `MBEDTLS_ERR_PK_INVALID_PUBKEY` for public keys.

RonEld commented on tests/data_files/public_rsa_key.pem:
```diff
@@ -0,0 +1,8 @@
+zB3Rm7u23xBBcLr+zb4fVjBOGS1vIVNnxj7aDYJTB9ZO2i+5MUch0BPHhsd3gf//
+u5ECyItnc+B50apbc/7wziwX1ABMvGVIWbvEMG68Vgst2kX91ojiDPZJej/c2xLR
+gpzage6SGEIQiCDQVIudSMnzZoltyMUmNwIDAQAB
+-----END RSA PUBLIC KEY-----
```
I looked in `tests/data_files/Makefile`, it only generates certificates from existing keys. I will rename the files though

hanno-arm replied:
@RonEld That is changing recently and includes keys, too.

RonEld replied:
I see, so actually, I think we shouldn't add the keys to the repository, but dynamically generate them when building

hanno-arm replied:
@RonEld With the large number of keys and the computational effort to generate them, I think that would be too big of an overhead. @mpg What's your opinion on that?

mpg replied:
Keys and certificates used for testing should be commited to the repo. Generating them dynamically would only increase complexity and slow down testing, for not real benefit (repo size is really not an issue). Despite its name, that Makefile is only documentation, not part of the build system.

RonEld replied:
Personally I don't like putting binaries in repositories, especially generated ones, as they can be problematic in future builds. But I see your point

RonEld commented:
@hanno-arm I addressed your comments and pushed a new commit. Please review. I found that PKCS#1 DER public key was not handled correct, I added it as well

hanno-arm commented:
@RonEld Could you elaborate on the bug you found and fixed?

RonEld commented:
@hanno-arm Basically, when I worked on commenting `tests/data_files/Makefile` I found that the key I thought was pkcs1 der was actually pkcs8 der format, and der format was actually not parsed for pkcs1. I added another call to `pk_parse_key_pkcs1_der` on the actual key buffer, in case error is `MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT`

RonEld commented:
@hanno-arm I addressed your comments, please review

hanno-arm commented on library/pkparse.c:
```diff
@@ -1253,14 +1253,59 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+
+    p =  (unsigned char *) key;
+    ret = pk_get_rsapubkey( &p, p + keylen, mbedtls_pk_rsa( *ctx ) );
+    if ( ret == 0 )
```
This way, a PEM encoded generic public key (guarded by `BEGIN/END PUBLIC KEY`) will not be immediately recognized as such, but instead will be passed to the DER-parsing routine `pk_get_rsapubkey`. While the latter will gracefully fail and lead to the key being subsequently correctly categorized, I think we shouldn't call any DER parsing routines on PEM encoded keys, as the latter can be easily recognized as such.

Concretely: Can we align the code with the structure that the private key parsing function `mbedtls_pk_parse_key`, i.e. go through all possible PEM formats first and only afterwards try the DER parsing routines?

RonEld replied:
I thought about doing that, but this adds more `#if defined(MBEDTLS_RSA_C)` which I am not too fond of adding too much.
In addition, even if we send the pkcs8 PEM formatted file to the pkcs1 der parsing, it will fail, and then we continue to do the PEM parsing.
In my opinion, it's more readable to first try pkcs1 parsing, and then do the pkcs8 parsing

hanno-arm replied:
My main argument is that I'd like to have `mbedtls_pk_parse_key` and `mbedtls_pk_parse_public_key` follow the same strategy, regardless of which one that is (although I prefer to do the PEM vs. DER distinction first) Let's wait for another opinion maybe?

RonEld replied:
I am starting to tend to agree with, to start with PEM, and then DER, since this is how `mbedtls_pk_parse_key` is implemented, but this means bigger change on the code. It means to modify(in pseudo code)
~~~
#if defined(MBEDTLS_PEM_PARSE_C)
if  parse PEM succeeds:
    do stuff
    parse DER
    return ret
#endif /* MBEDTLS_PEM_PARSE_C */
parse DER
~~~
Following this, I will also need to change the regular parsing, which might be a bit out of scope of this PR

RonEld commented:
@hanno-arm I addressed your comments. I also found issues when `MBEDTLS_PEM_PARSE_C` not defined, and fixed them as well.
Please review

mpg commented on library/pkparse.c:
```diff
@@ -1250,17 +1250,49 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+
+        if ( ( ret = pk_get_rsapubkey( &p, p + pem.buflen, mbedtls_pk_rsa( *ctx ) ) ) != 0 )
+            mbedtls_pk_free( ctx );
+        mbedtls_pem_free( &pem );
```
Style: please leave an empty line between the preceding block and this one.

mpg commented on library/pkparse.c:
```diff
@@ -1250,17 +1250,49 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+    }
+#endif /* MBEDTLS_RSA_C */
+
+       /* Avoid calling mbedtls_pem_read_buffer() on non-null-terminated string */
```
Style: please use the same indentation level as the following line of code.

mpg commented on tests/data_files/Makefile:
```diff
@@ -64,7 +64,13 @@ server2-sha256.crt: server2-rsa.csr
+all_final += rsa_pkcs8_2048_public.pem
 
+rsa_pkcs1_2048_public.der: rsa_pkcs1_2048_public.pem
+	$(OPENSSL) -RSAPublicKey_in -in rsa_pkcs1_2048_public.pem -outform DER -RSAPublicKey_out -out $@
```
Typo: missing `rsa` command.

(Also, might want to use `$<` instead of repeating the input file name, here and above.)

mpg commented on tests/data_files/Makefile:
```diff
@@ -64,7 +64,13 @@ server2-sha256.crt: server2-rsa.csr
 
+rsa_pkcs1_2048_public.der: rsa_pkcs1_2048_public.pem
+	$(OPENSSL) -RSAPublicKey_in -in rsa_pkcs1_2048_public.pem -outform DER -RSAPublicKey_out -out $@
+all_final += rsa_pkcs8_2048_public.der
```
Typo: you probably mean `pkcs1` (not `8`) here.

Also, the command for making the `pkcs8`-wrapped version seem to be missing from the Makefile.

RonEld replied:
I didn't write the `pkcs8` one, as I saw a missing test, and I only converted the existing `PEM` file. 
I'll add both `pkcs8` files, as I will create a new one, based on SHA256 instead of MD5

mpg commented on tests/suites/test_suite_pkparse.data:
```diff
@@ -106,6 +106,18 @@ Parse Public RSA Key #1 (PKCS#8 wrapped)
 pk_parse_public_keyfile_rsa:"data_files/format_gen.pub":0
 
+Parse Public RSA Key #1 (PKCS#8 wrapped, DER)
+depends_on:MBEDTLS_MD5_C
```
Can you create a test case that uses SHA-256 instead (either in replacement or in addition to that one)? So that modern configurations that don't have MD5 or SHA1 still have a test for that feature.

RonEld replied:
sure, I converted the original `format_gen.pub` file to `pkcs1` , but I will create a new pair of `pkcs8` `PEM` and `DER`

RonEld replied:
@mpg I have just looked in the code, and code itself requires PEM format encrypted only when `MBEDTLS_MD5_C` is defined, which is strange to me why there is a need for an `MD` algorithm for encrypted `PEM` in the first place..

RonEld replied:
Nevermind, it uses `pem_pbkdf1` internally, which uses `MD5`

hanno-arm replied:
@RonEld That's because there's a PBKDF in between the PEM password and the actual encryption key used (it's a deliberately expensive function to slow down dictionary attacks).

RonEld replied:
@hanno-arm yes, I saw it after posting.

RonEld commented:
@mpg Thank you for your review!
I addressed your comments, and changed accordingly

RonEld commented:
@hanno-arm Are you happy with the latest result?

Keno commented:
Bump. I ran into this issue as well. Would be lovely to have this fixed.

hanno-arm commented on ChangeLog:
```diff
@@ -4,6 +4,7 @@ mbed TLS ChangeLog (Sorted per branch, date)
 
 Features
    * Allow comments in test data files.
+   * Add support for public keys encoded in PKCS#1 format
```
Style niggle: Could you please add a dot at the end?

RonEld replied:
done

hanno-arm commented on library/pkparse.c:
```diff
@@ -1270,23 +1304,43 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+    mbedtls_pk_free( ctx );
+    if ( ret != ( MBEDTLS_ERR_PK_INVALID_PUBKEY + MBEDTLS_ERR_ASN1_UNEXPECTED_TAG ) )
+    {
+        return ( ret );
```
Minor: The spacing is not correct here.

hanno-arm commented on library/pkparse.c:
```diff
@@ -1270,23 +1304,43 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+    if( ( ret = mbedtls_pk_setup( ctx, pk_info ) ) != 0 )
+        return( ret );
+
+    p =  (unsigned char *) key;
```
Minor: The spacing is not correct here.

hanno-arm commented on library/pkparse.c:
```diff
@@ -1270,23 +1304,43 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+
+    p =  (unsigned char *) key;
+    ret = pk_get_rsapubkey( &p, p + keylen, mbedtls_pk_rsa( *ctx ) );
+    if ( ret == 0 )
```
Spacing again (no space after `if`).

hanno-arm commented on library/pkparse.c:
```diff
@@ -1270,23 +1304,43 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+        return( ret );
+    }
+    mbedtls_pk_free( ctx );
+    if ( ret != ( MBEDTLS_ERR_PK_INVALID_PUBKEY + MBEDTLS_ERR_ASN1_UNEXPECTED_TAG ) )
```
Minor: No space after `if`.

hanno-arm commented on tests/suites/test_suite_pkparse.data:
```diff
@@ -103,8 +103,19 @@ depends_on:MBEDTLS_DES_C:MBEDTLS_SHA1_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_PKCS5_C:MBED
+pk_parse_public_keyfile_rsa:"data_files/rsa_pkcs8_2048_public.pem":0
+
+Parse Public RSA Key #1 (PKCS#8 wrapped, DER)
+pk_parse_public_keyfile_rsa:"data_files/rsa_pkcs8_2048_public.der":0
```
The dependency on `MBEDTLS_RSA_C` seems to be missing here.

RonEld replied:
no need, because [`pk_parse_public_keyfile_rsa`](https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_pkparse.function#L12) tests case is dependent on `MBEDTLS_RSA_C`

hanno-arm commented on tests/suites/test_suite_pkparse.data:
```diff
@@ -103,8 +103,19 @@ depends_on:MBEDTLS_DES_C:MBEDTLS_SHA1_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_PKCS5_C:MBED
-depends_on:MBEDTLS_MD5_C:MBEDTLS_PEM_PARSE_C
-pk_parse_public_keyfile_rsa:"data_files/format_gen.pub":0
+depends_on:MBEDTLS_PEM_PARSE_C
+pk_parse_public_keyfile_rsa:"data_files/rsa_pkcs8_2048_public.pem":0
```
The dependency on `MBEDTLS_RSA_C` seems to be missing here.

RonEld replied:
no need, because [`pk_parse_public_keyfile_rsa`](https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_pkparse.function#L12) tests case is dependent on `MBEDTLS_RSA_C`

hanno-arm replied:
Thanks!

RonEld commented:
@hanno-arm I addressed your issues, and removed the dependency on `MBEDTLS_RSA_C` in the additional tests that did add this dependency, even though the test is dependent on this configuration.

hanno-arm commented on library/pkparse.c:
```diff
@@ -1250,11 +1250,45 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
+    {
+        mbedtls_pem_free( &pem );
+        return( ret );
+    }
```
There is a slight abstraction break happening here (without consequences): If we fall through the RSA PEM case, then `pem` is no longer guaranteed by the API to be in initialized state (it has already been used in `mbedtls_pem_read_buffer` after `mbedtls_pem_init`), and hence there should be a call to `mbedtls_pem_free` and `mbedtls_pem_init` before the next call to `mbedtls_pem_read_buffer`. @RonEld Could you add such a pair at the end of the `MBEDTLS_RSA_C` block?

RonEld replied:
@hanno-arm I may be not following you. This is same logic as in `mbedtls_pk_parse_key` , isn't it?

hanno-arm replied:
@RonEld The point is that `pem` should be in a freshly initialized state when we fall through to the next call of `mbedtls_pem_read_buffer` a few lines below -- but, at least as far as the API is concerned, that's not guaranteed, because we have already called `mbedtls_pem_read_buffer` once in between. Calling `mbedtls_pem_free( &pem ); mbedtls_pem_init( &pem );` at the end of the `MBEDTLS_RSA_C` block would fix that.

hanno-arm replied:
@RonEld You're right, the private key parsing functions do it in the same not totally correct, yet non-critical, way. So let's keep it for now.

mpg commented:
The CI failures are due to know issues that have been fixed in development since that branch diverged, and are in any case independent of the changes in that branch - so OK to merge.

This only extends functionality of existing functions but does not change the public API, so I believe no architectural approval is needed, therefore I'm tagging as ready to merge.

gilles-peskine-arm commented:
CI failed due to issues (timing tests, IAR8) that have been fixed in test scripts in development. Re-running on a merge with development: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/89/
